src/actions/profile-view.js

```diff
@@ -73,86 +78,364 @@ export function setCallNodeContextMenuVisibility(isVisible: boolean): Action {
-    const threadOrder = getThreadOrder(getState());
-    const hiddenThreads = getHiddenThreads(getState());
+    const globalTrackToHide = getGlobalTracks(getState())[trackIndex];
+    let selectedThreadIndex = getSelectedThreadIndex(getState());
+
+    // Re-select the selectedThreadIndex if it is hidden with this operation.
+    if (globalTrackToHide.type === 'process') {
+      // This is a process global track, this operation could potentially hide
+      // the selectedThreadIndex.
+      let isSelectedThreadIndexHidden =
+        globalTrackToHide.mainThreadIndex === selectedThreadIndex;
+
+      // Check in the local tracks for the selectedThreadIndex
+      if (!isSelectedThreadIndexHidden) {
+        for (const localTrack of getLocalTracks(
+          getState(),
+          globalTrackToHide.pid
+        )) {
+          if (
+            localTrack.type === 'thread' &&
+            localTrack.threadIndex === selectedThreadIndex
+          ) {
+            isSelectedThreadIndexHidden = true;
+            break;
+          }
+        }
+      }
+      if (isSelectedThreadIndexHidden) {
+        selectedThreadIndex = _findOtherVisibleThread(getState, trackIndex);
+      }
+    }
 
-    // Do not allow hiding the last thread.
-    if (hiddenThreads.length + 1 === threadOrder.length) {
+    if (selectedThreadIndex === null) {
+      // Hiding this process would make it so that there is no selected thread.
+      // Bail out.
       return;
     }
 
-    const threads = getThreads(getState());
-    const thread = threads[threadIndex];
     sendAnalytics({
       hitType: 'event',
-      eventCategory: 'threads',
-      eventAction: 'hide',
-      eventLabel: getFriendlyThreadName(threads, thread),
+      eventCategory: 'timeline',
+      eventAction: 'hide global track',
     });
 
-    dispatch(
-      ({
-        type: 'HIDE_THREAD',
-        threadIndex,
-        threadOrder,
-        hiddenThreads,
-      }: Action)
-    );
+    dispatch({
+      type: 'HIDE_GLOBAL_TRACK',
+      trackIndex,
+      selectedThreadIndex,
+    });
+  };
+}
+
+export function showGlobalTrack(trackIndex: TrackIndex): ThunkAction<void> {
+  return dispatch => {
+    sendAnalytics({
+      hitType: 'event',
+      eventCategory: 'timeline',
+      eventAction: 'show global track',
+    });
+
+    dispatch({
+      type: 'SHOW_GLOBAL_TRACK',
+      trackIndex,
+    });
   };
 }
 
-export function showThread(threadIndex: ThreadIndex): ThunkAction<void> {
+export function isolateGlobalTrack(
+  isolatedTrackIndex: TrackIndex
+): ThunkAction<void> {
   return (dispatch, getState) => {
-    const threads = getThreads(getState());
-    const thread = threads[threadIndex];
+    const track = getGlobalTracks(getState())[isolatedTrackIndex];
+    const trackIndexes = getGlobalTrackOrder(getState());
+
+    if (track.type !== 'process') {
+      // Do not isolate a track unless it is a process, that way a thread
+      // will always be visible.
+      return;
+    }
+
```

Comment:
I'm not sure exactly what you mean here. When isolating a global process, it will ensure that there is a valid `selectedThreadIndex`, and if it's the same one that is currently selected then it's a noop.

Reply:
This is the example I'm thinking of:
```
GlobalTrack A
  - LocalTrack 1
  - LocalTrack 2 <selected>
GlobalTrack B
  - bla bla bla
```

Isolating `GlobalTrack A` should keep `LocalTrack 2` selected. But as I understand from your code is that you'll try to select a new thread index as if nothing was selected. This could make `Global Track A` selected if possible, or otherwise `LocalTrack 1`. I think this behavior would be confusing.

I hope it's clearer !

Reply:
Ah, seems reasonable, I'll tweak it to do that.